### ISSUE TITLE
High memory usage

### ISSUE THREAD

**Issue by mxmlnkn** [Rogdham/python-xz#2]
Hello,

During running some of my benchmarks for ratarmount, I noticed one of them failing with a 100GB .tar.xz test file. The memory usage increases while linearly while reading over the file and exceeded 33GB and which point it was killed by the OOM killer. It works with other compression backends, so I'm pretty sure the memory usage stems from python-xz. Do you have any idea from where it might originate from? I would have thought that you only need to story block and stream offsets, i.e., list of integers, which should be small.

**Comment by mxmlnkn** [comment]
Some code to reproduce the problem:

```python3
import resource
import sys
import time

import xz


if __name__ == '__main__':
    print("xz version:", xz.__version__)

    size = 0
    t0 = time.time()
    with xz.open(sys.argv[1], 'rb') as file:
        t1 = time.time()

        while True:
            readSize = len(file.read(32*1024*1024))
            if readSize == 0:
                break
            size += readSize

            if time.time() - t1 > 10:
                t1 = time.time()
                print(f"{t1 - t0:.0f}s {resource.getrusage(resource.RUSAGE_SELF).ru_maxrss // 1024} MiB RSS")

    t1 = time.time()
    print(f"Reading {size} B took: {t1-t0}s")

"""
base64 /dev/urandom | head -c $(( 4*1024*1024*1024  )) > large
xz -T 0 --keep large
xz -l large.xz
    Strms  Blocks   Compressed Uncompressed  Ratio  Check   Filename
        1     171  3.148,1 MiB  4.096,0 MiB  0,769  CRC64   large.xz
python3 decompress-xz.py large.xz
```

Output:

    xz version: 0.2.0
    10s 171 MiB RSS
    21s 243 MiB RSS
    31s 323 MiB RSS
    41s 396 MiB RSS
    51s 468 MiB RSS
    61s 548 MiB RSS
    71s 621 MiB RSS
    82s 693 MiB RSS
    92s 773 MiB RSS
    102s 846 MiB RSS
    112s 918 MiB RSS
    123s 998 MiB RSS
    133s 1071 MiB RSS
    143s 1143 MiB RSS
    153s 1223 MiB RSS
    163s 1296 MiB RSS
    174s 1368 MiB RSS
    184s 1448 MiB RSS
    Reading 4294967296 B took: 186.66424465179443s

I also tested 0.3.1 and it behaves the same.

**Comment by Rogdham** [comment]
Thank you for your report and the test case, I'm able to reproduce.

After some profiling, I can see that the LZMA decompressors (1 per block) are responsible for all that memory usage.

I see a quick but partial fix (case 1), and a more difficult fix (case2).

---

### Case 1

When a LZMA block is fully read (i.e. last byte of that block is read), free the LZMA decompressor. To be honest, I'm surprised Python was not already doing it automatically.

This should have no side effect, and fix your test example.


### Case 2

However, if partial reads are performed from many blocks, the issue would remain.

Imagine the following scenario:
 - read 1kb from block A
 - jump to block B, read 1kb
 - jump to where we were in block A, read next kb
 - jump to where we were in block B, read next kb
 - jump to where we were in block A, read next kb
 - etc.

In this scenario (continuous reads but alternating from two blocks), we would want the LZMA decompressors to stay in memory, otherwise reads will be done from the beginning of a block each time we switch block (which would be slower).

Imagine the same scenario with more than 2 alternating blocks.

**We have a tradeoff between memory usage and speed to resume reading from the middle of a block.**

Maybe I could draw a line at a an arbitrary number of LZMA decompressors (e.g. 8, but this would be parametrizable) and make sure that we delete LZMA decompressors (we would need to decide which ones to delete) so that no more than 8 are alive at any point in time.

---

My plan is to work on the first case by this weekend, and think about the second case after.

@mxmlnkn I'm afraid real-world use of ratarmount would hit the case 2, but maybe I'm wrong?

Also, I would appreciate any feedback for the case 2!

**Comment by mxmlnkn** [comment]
I think you are right. The second case isn't that abnormal and might happen with ratarmount depending on the usage of the mount point.

Maybe you can implement a kind of block cache to handle case 2 and still be memory-bound and fast in most cases. I.e., hold the last 10 (or any number) most recently used blocks and deallocate the LzmaDecompressor of any older one even if the end of the block was not reached.

**Comment by mxmlnkn** [comment]
I still don't understand while neither closing nor deleting the opened XZ file, doesn't free the memory in my test. I would have expected the memory to be freed after the with statement or at least after `del file`, which both don't seem to improve anything in my tests.

**Comment by Rogdham** [comment]
> Maybe you can implement a kind of block cache to handle case 2 and still be memory-bound and fast in most cases. I.e., hold the last 10 (or any number) most recently used blocks and deallocate the LzmaDecompressor of any older one even if the end of the block was not reached.

This was a proposition I tried to make in my last message, but it may have not be written in the best way. In any case, if we both think that's a possible solution, it must be a good idea :wink: 

---

> I still don't understand while neither closing nor deleting the opened XZ file, doesn't free the memory in my test. I would have expected the memory to be freed after the with statement or at least after del file, which both don't seem to improve anything in my tests.

From my understanding, `ru_maxrss` gives the **peak** memory usage of the process, and in many cases will never decrease: https://unix.stackexchange.com/a/30941

Btw because of this writing tests targeting memory consumption is kind of hard :unamused: 

---

> Case 2 is not even about performance, it also still makes it possible to explode memory until the out-of-memory limit has been hit if all xz blocks are only partially read

yes exactly

**Comment by mxmlnkn** [comment]
> This was a proposition I tried to make in my last message, but it may have not be written in the best way. In any case, if we both think that's a possible solution, it must be a good idea wink

Ah, I see now. Yes, that seems like the way to go for me albeit it is complex.

> From my understanding, `ru_maxrss` gives the **peak** memory usage of the process, and in many cases will never decrease: https://unix.stackexchange.com/a/30941

That is also how I understand it. But I **am** interested in the peak memory usage because that is what determines whether I am able to run the program or not with limited main memory. Even if it requires 500GB of RAM for only a second that still would lead to an out of memory kill (or hanging Linux if the OOM killer does not work).

**Comment by Rogdham** [comment]
I've pushed some commits that provide a first implementation. Please note that I'm not sure about the exposed API of providing a custom strategy for freeing LZMA decompressors, and I have tests against memory requirements failing in some Python versions.

But the fix should be working as it is without any change needed in your code, could you try and see if it solves the issue on your 100Gb file?

```
pip install git+https://github.com/Rogdham/python-xz.git@68bb6a6b81de9a14ad67d3d59122c7307f0aab8b
```

**Comment by mxmlnkn** [comment]
I tried it with a smaller file for now but it seems to work. Thank you!

An API for a custom strategy for freeing looks good but also "difficult". Imho, the default version should behave sufficiently well, i.e. not run out of memory in any case, without any tuning necessary by the library user. But by itself, the strategy classes look good. If an abstract base class is provided, a developer could even write his own strategies as he requires, but that would definitely be something for power users. E.g., one strategy right off my mind would be "clear decompressors if free memory is below a threshold".

**Comment by Rogdham** [comment]
Thank you for all your feedback!

Fixed in version 0.4.0 :tada:
